processor/labelsprocessor/labels_processor.go

```diff
@@ -0,0 +1,107 @@
+package labelsprocessor
+
+import (
+	"context"
+	"fmt"
+
+	"go.opentelemetry.io/collector/consumer/pdata"
+	v11 "go.opentelemetry.io/collector/internal/data/opentelemetry-proto-gen/common/v1"
+	v1 "go.opentelemetry.io/collector/internal/data/opentelemetry-proto-gen/metrics/v1"
+)
+
+type labelMetricProcessor struct {
+	cfg *Config
+}
+
+func newLabelMetricProcessor(cfg *Config) (*labelMetricProcessor, error) {
+	err := validateConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+	return &labelMetricProcessor{cfg: cfg}, nil
+}
+
+func validateConfig(cfg *Config) error {
+	// Ensure no empty keys/values exist
+	for _, elem := range cfg.Labels {
+		if elem.Key == "" || elem.Value == "" {
+			return fmt.Errorf("Labels Processor configuration contains an empty key or value")
+		}
+	}
+
+	//Ensure no duplicate keys exist
+	keys := make(map[string]bool)
+	for _, elem := range cfg.Labels {
+		_, value := keys[elem.Key]
+		if value {
+			return fmt.Errorf("Labels Processor configuration contains duplicate keys")
+		}
+		keys[elem.Key] = true
+	}
+
+	return nil
+}
+
+func (lp *labelMetricProcessor) ProcessMetrics(_ context.Context, md pdata.Metrics) (pdata.Metrics, error) {
+
+	otlpMetrics := pdata.MetricsToOtlp(md)
+
+	for _, otlpMetric := range otlpMetrics {
+		for _, instrMetric := range otlpMetric.GetInstrumentationLibraryMetrics() {
+			for _, metric := range instrMetric.GetMetrics() {
+
+				// Multiple types of Data Points exists, since there is no way to determine this beforehand we initialize variables for all
+				var intDataPoint []*v1.IntDataPoint
+				var doubleDataPoint []*v1.DoubleDataPoint
+				var intHistogramDataPoint []*v1.IntHistogramDataPoint
+				var doubleHistogramDataPoint []*v1.DoubleHistogramDataPoint
+
+				if metric.GetDoubleGauge() != nil {
+					doubleDataPoint = metric.GetDoubleGauge().GetDataPoints()
+				} else if metric.GetDoubleHistogram() != nil {
+					doubleHistogramDataPoint = metric.GetDoubleHistogram().GetDataPoints()
+				} else if metric.GetDoubleSum() != nil {
+					doubleDataPoint = metric.GetDoubleSum().GetDataPoints()
+				} else if metric.GetIntGauge() != nil {
+					intDataPoint = metric.GetIntGauge().GetDataPoints()
+				} else if metric.GetIntHistogram() != nil {
+					intHistogramDataPoint = metric.GetIntHistogram().GetDataPoints()
+				} else if metric.GetIntSum() != nil {
+					intDataPoint = metric.GetIntSum().GetDataPoints()
+				}
```

Comment:
I believe each `metric` has a determined type that you can use to determine which type of datapoint it has. Not sure if that would be better than trying to see if those types of datapoints are null. 

Also, you might be able to bring this logic into the next for loop which I think might be a big cleaner? Just so that you don't have to explain only 1 of the 4 for loops that you have will run.

Reply:
Refactoring of this was discussed in private. Instead of the seperate paths from the if statements converging here, each if statement will handle its data type within itself